rust/kernel/cpu.rs

```diff
@@ -0,0 +1,131 @@
+    /// Constructs a new instance referring to a specific processor by id.
+    ///
+    /// # Safety
+    ///
+    /// This function checks that the passed id is a valid _potential_ processor id, but does
+    /// not check that the processor is actually present or online. Manipulation of absent
+    /// or offline processors will fail with reasonable errors returned.
+    pub fn new(id: i32) -> Result<Cpu> {
+        if id >= Self::num_possible() {
+            return Err(code::E2BIG);
+        } else if id < 0 {
+            return Err(code::EINVAL);
+        }
+        Ok(Cpu(id))
+    }
+
+    /// Returns the number of processors that are possible (including onlining and hotplug).
+    pub fn num_possible() -> i32 {
+        // SAFETY: The number of possible CPUs is fixed early at boot and does not change in
+        // response to processors coming online, offline, or hotplug events.
+        unsafe { bindings::nr_cpu_ids as i32 }
+    }
+
+    /// Lock the running thread to the current CPU and prevent preemption.
+    pub fn lock_current() -> Guard {
+        Guard::new()
+    }
+
+    /// Get the processor id of this CPU.
+    pub fn id(&self) -> i32 {
+        self.0
+    }
+
+    /// Call a function on the CPU identified by `self`.
+    pub fn call<R, F>(&self, func: F) -> R
+    where
+        F: FnOnce() -> R + Send,
+        R: Send,
+    {
+        // Evaluate func locally if we're already on the right CPU.
+        {
+            let guard = Self::lock_current();
+            if self.0 == guard.cpu().id() {
+                return func();
+            }
+        }
+
+        let mut fn_call = FnCall {
+            out: None,
+            func: ManuallyDrop::new(func),
+        };
+        // SAFETY: smp_call_function_single is invoked with wait == 1 (true) ensuring that fn_call
+        // outlives this FFI invocation. func's returned value is moved into fn_call.out and then
+        // returned here. The Send type constraint ensures func and its returned value can safely
+        // be moved between threads/CPUs.
+        unsafe {
+            bindings::smp_call_function_single(
+                self.0,
+                Some(call_fn::<F, R>),
+                (&mut fn_call as *mut FnCall<F, R>).cast(),
+                1,
+            );
+        }
+        fn_call.out.unwrap()
+    }
+}
+
+/// CPU guard preventing preemption of the current thread and keeping it locked to a single CPU.
+pub struct Guard(
+    // Id of the CPU this guard is locked to
+    i32,
+    // Wrap a pointer type to inhibit Rust's default Sync and Send trait impls.
+    PhantomData<*mut ()>,
+);
+
+impl Guard {
+    /// Lock the current thread to the current processor.
+    fn new() -> Self {
+        // SAFETY: FFI invocation; matching put_cpu occurs on `drop()`. Guard is neither Sync nor Send.
+        let id = unsafe { bindings::get_cpu() };
+        Self(id, PhantomData)
+    }
+
+    /// Obtain a Cpu referring to the current (locked) processor.
```

Comment:
This text gets extracted to html, by convention we want to cross-reference types, so it would be better to have use ``[`Cpu`]`` here so that the extracted text will have a link to `Cpu`.

Reply:
Done.